### PR TITLE
Feature/auto pause fundraising

### LINKED ISSUE #202: implement auto-pause for ContinuousFundraising
Regulation allows fundraising with an end date. Our contracts should respect that and auto-pause themselves after the end date.

### PR DESCRIPTION
**Pull request by malteish**
closes #202

### CODE REVIEW

File: contracts/PublicFundraising.sol [hunk 1]
```diff
@@ -259,12 +274,21 @@ contract PublicFundraising is
         coolDownStart = block.timestamp;
     }
 
+    /// set auto pause date
+    function setAutoPauseDate(uint256 _autoPauseDate) external onlyOwner whenPaused {
+        autoPauseDate = _autoPauseDate;
+    }
+
+    function _pause() internal override(PausableUpgradeable) {
```
Comment (CJentzsch) [inline contracts/PublicFundraising.sol:282]: do we really need that function? Can we not just use the "pause" function?
Or make this function external and remove the other pause function?

**Review by CJentzsch — CHANGES_REQUESTED** [review]

File: contracts/PublicFundraising.sol (hunk 1 above)
Comment (malteish) [reply contracts/PublicFundraising.sol:282]: I agree it does not make a huge difference in code quality. But then, does not make a relevant difference in gas cost either:
```
Ran 1 test suites: 46 tests passed, 0 failed, 0 skipped (46 total tests)
testAutoPauseInConstructor(uint256,uint256) (gas: 0 (0.000%)) 
testBuyAndMintToDifferentAddress() (gas: 0 (0.000%)) 
testBuyHappyCase(uint256) (gas: 0 (0.000%)) 
testBuySmallAmountAfterInitialInvestment() (gas: 0 (0.000%)) 
testBuyTooLittle() (gas: 0 (0.000%)) 
testBuyTooMuch() (gas: 0 (0.000%)) 
testConstructorHappyCase() (gas: 0 (0.000%)) 
testConstructorWithAddress0() (gas: 0 (0.000%)) 
testCorrectAccounting() (gas: 0 (0.000%)) 
testExceedMintingAllowance() (gas: 0 (0.000%)) 
testFailOverflow() (gas: 0 (0.000%)) 
testFailUnderflow() (gas: 0 (0.000%)) 
testFailUpdateCurrencyAndPriceNotPaused() (gas: 0 (0.000%)) 
testFailUpdateCurrencyReceiverNotPaused() (gas: 0 (0.000%)) 
testFailUpdateMaxAmountOfTokenToBeSoldNotPaused() (gas: 0 (0.000%)) 
testFailUpdateMaxAmountPerBuyerNotPaused() (gas: 0 (0.000%)) 
testFailUpdateMinAmountPerBuyerNotPaused() (gas: 0 (0.000%)) 
testMultipleAddressesBuyForOneReceiver() (gas: 0 (0.000%)) 
testMultiplePeopleBuyTooMuch() (gas: 0 (0.000%)) 
testReentrancy() (gas: 0 (0.000%)) 
testRevertsOnOverflow(uint256,uint256) (gas: 0 (0.000%)) 
testRoundsUp(uint256,uint256) (gas: 0 (0.000%)) 
testTransferOwnership(address) (gas: 0 (0.000%)) 
testUpdateCurrencyAndPricePaused(uint256) (gas: 24 (0.002%)) 
testBuyAnyAmountRoundsUp(uint256) (gas: 24 (0.007%)) 
testBuy1BitRoundsUp() (gas: 24 (0.008%)) 
testUnpauseTooSoonAfterSetCurrencyAndTokenPrice(uint128,uint32,uint32,uint256) (gas: 24 (0.028%)) 
testUnpauseTooSoonAfterSetMinAmountPerBuyer(uint128,uint32,uint32,uint256) (gas: 24 (0.028%)) 
testUnpauseTooSoonAfterSetMaxAmountPerBuyer(uint128,uint32,uint32,uint256) (gas: 24 (0.028%)) 
testUpdateMinAmountPerBuyerPaused(uint256) (gas: 24 (0.029%)) 
testUnpauseTooSoonAfterSetCurrencyReceiver(uint128,uint32,uint32,address) (gas: 24 (0.029%)) 
testUnpauseTooSoonAfterSetMaxAmountOfTokenToBeSold(uint128,uint32,uint32) (gas: 24 (0.030%)) 
testLogicContractCreation() (gas: 600 (0.030%)) 
testUpdateMaxAmountPerBuyerPaused(uint256) (gas: 24 (0.031%)) 
testUpdateCurrencyReceiverPaused() (gas: 24 (0.033%)) 
testUnpauseAfterSetMinAmountPerBuyer() (gas: 19 (0.033%)) 
testUpdateMaxAmountOfTokenToBeSoldPaused(uint256) (gas: 24 (0.033%)) 
testUnpauseAfterSetCurrencyReceiver() (gas: 19 (0.034%)) 
testUnpauseAfterSetMaxAmountOfTokenToBeSold() (gas: 19 (0.034%)) 
testAutoPauseActivation(uint256,uint256) (gas: 48 (0.034%)) 
testUnpauseAfterSetMaxAmountPerBuyer() (gas: 20 (0.034%)) 
testFailUnpauseAfterDelay() (gas: 24 (0.037%)) 
testFailUnpauseImmediatelyAfterPausing() (gas: 24 (0.038%)) 
testUnpauseAfterDelayAnd1Sec() (gas: 19 (0.039%)) 
testUnpauseAfterSetCurrencyAndTokenPrice() (gas: 24 (0.039%)) 
testFailPaused() (gas: 24 (0.040%)) 
Overall gas change: 1128 (0.006%)
```
So I would like to keep it.

Comment (malteish) [reply contracts/PublicFundraising.sol:282]: We could ask if the auto-pause function should even activate the cool down period. If no parameters are changed, why enforce a cool down? And once we go down that road, why enforce a cool down on manual pause activation? Pause does not have to update `coolDownStart`, it does not need a dedicated internal function either.

Comment (CJentzsch) [reply contracts/PublicFundraising.sol:338]: If not needed, less code is always better.

**Review by CJentzsch — APPROVED** [review]

**Pull request merged**